README
##############################################################################
COPYRIGHT: University of California Berkeley 2006-2007
EMAIL: support@cirl.berkeley.edu

Recon Tools:

A collection of tools and libraries for reconstructing and undistorting MR 
image data developed at the University of California Berkeley's Helen Wills 
Neuroscience Institute Brain Imaging Center.
For documentation, see https://cirl.berkeley.edu/view/BIC/ImagingTools

#-----------------------------------------------------------------------------
SUMMARY:

The Recon Tools package provides a number of command-line tools for
manipulating MR image data, as well as a code library of reusable components.
The central types of objects in the recon-tools system are data readers,
transformers, and writers.  Image manipulation tools can be created by
composing appropriate combinations of these core objects into a pipeline
beginning with a reader, followed by a sequence of transformers (called
Operations), and ending with a writer.  Image data in the pipeline is
represented by an Image object conforming to a conventional Image interface.

#-----------------------------------------------------------------------------
LIBRARY DEPENDENCIES:
Recon Tools is written in Python 2.4, making use these additional libraries:

numpy > 1.0
matplotlib > 0.87

#-----------------------------------------------------------------------------
BUILDING/INSTALLING THE PYTHON WAY
It is necessary to compile the phase surface unwrapping module included 
with Recon Tools before installing. To do so, type:

python setup.py build
sudo python setup.py install

#-----------------------------------------------------------------------------
USING MAKE (the iffy way):

make install:       installs tools, library, and documentation
make tools-rpm:     build an RPM for installing the tools and library
make docs:          generate some documentation from code
make install-docs:  install documentation (/usr/share/doc/recon-tools)
make docs-rpm:      build an RPM for install the documentation
make test:          run test suite
make clean:         remove all generated files

#-----------------------------------------------------------------------------
TESTING:

To test your installation, a collection of example datasets and oplists is 
available at the CIRL wiki:

https://cirl.berkeley.edu/twiki/pub/BIC/ReconTools/testsuite.tar.bz2

To use the testing util, type:

tar jxf testsuite.tar.bz2
cd testsuite
./test.sh



release-notes
recon-tools-0.?????                                                  ????????
- more speed ups from compiled routines
- IMPORTANT PLEASE READ!
  under a more literal interpretation of Varian data, ReorderSlices no longer
  changes the slicing direction. For example, in axial slicing, the slicing
  axis runs negative to positive from the top of the head towards the feet.
  Whereas ReorderSlices used to reflect that axis, so that slicing ran feet
  to head, it no longer does! However, when changing the orientation to some
  canonical form (eg radiological for the axial plane), the slicing direction
  WILL potentially be reflected in accordance with the standard.
------------------------------------------------------------------------------
recon-tools-0.5-r5                                                 05/08/2008
- new "recon_gui" tool allows users to prepare and test oplists in a 
  graphical environment. Type "recon_gui <filename>" at the command line.
- yes, some bug fixes regarding image orientation when using RotPlane
  followed by FlipSlices (LR flip info did not get carried through!!)
- SliceView tool can now save movies of time-series or slices (requires
  mencoder program, part of mplayer software)
- new numpy style code tests
- revisited TimeInterpolate operation to make it useful again
- some memory leak fixes in unwrapping extension code
- VERY EXPERIMENTAL/VOLATILE support for Siemens N-channel raw data

------------------------------------------------------------------------------
recon-tools-0.5-r4  (beta)                                         10/03/2007
- bugfix in FixTimeSkew necessitated re-release of code

------------------------------------------------------------------------------
recon-tools-0.5-r3  (beta)                                         09/14/2007
- many speedups, yielding ~ 30% reduction in processing time
- UnbalPhaseCorrection has been re-factored to use more data in the model
  fitting. Most importantly, the parameters have changed: there is now a
  single parameter called "percentile" (default 75.0). This controls how
  many of the higher quality point will be used.
- New 3D phase unwrapping method, implemented in fieldmap calculations. This
  appears to be much more robust than the former 3D unwrapping
- Improvements to the Sliceview util
- A new orthogonal plotter a la SPM
- There is a new computational library dependency: FFTW3 compiled in double
  precision and single precision mode.

------------------------------------------------------------------------------
recon-tools-0.5-r2  (beta)                                         06/26/2007
- a new BalPhaseCorrection routine is available
- several new features in sliceviewer, including image overlays
- new option -e, --opsexample in main recon script: if you're not sure how
  to reconstruct an image, type "recon -e some-data.fid" to print 
  an example oplist to the screen
- NIFTI orientation descriptions in RotPlane fixed, also fixed "recon_epi"
  mode, which is equivalent to neurological orientation in the axial plane.
  (You no longer need to follow RotPlane by FlipSlices.)
- UnbalPhaseCorrection "thresh" parameter is 0.1 by default (was 1.0)

------------------------------------------------------------------------------
recon-tools-0.5-r1  (beta)                                         02/26/2007
- switching numerical routines to use numpy library for speedup
- major overhaul of image class
- RotPlane operation replaces Rot90, which never really worked as
  advertised--see RotPlane wiki for more details on usage
   * https://cirl.berkeley.edu/view/BIC/RotPlane
- flip_slices option removed from ReorderSlices and FixTimeSkew, due to
  conflicting action in RotPlane
- NIFTI and ANALYZE orientation/transformation fields improved
- more output datatypes are supported with -y flag; see recon --help
- new -u command line flag for online op-usage: recon -u OpName
- stability fix for ref-vol phase unwrapping, works better in brains

------------------------------------------------------------------------------
recon-tools-0.4-r2  (beta)                                         01/16/2007
- small bugfixes regarding:
   * mpflash slice-direction scaling
   * ComputeFieldMap phase unwrapping
   * FixTimeSkew on an odd-number of data points
- added option to tweak the asems-signal threshold in ComputeFieldMap
- added multiple colormap options to the slice viewer

This will be the last version of Recon Tools to run with the aging Numeric
library. The upcoming Recon Tools will use the "numpy" library, and will
require a migration to numpy and matplotlib 87.7 (or any matplotlib configured
for use with numpy).

------------------------------------------------------------------------------
recon-tools-0.4     (beta)                                         11/??/2006

  ***Libraries have changed location/name reflecting project name change!
     The "imaging" python library location is now known as "recon".
     Admins should delete the imaging/ directory from "site-packages" ***

- New features in the slice viewer:
    *zoom
    *save-as-png
    *save-montage-as-png
    *contour plotting
- FixTimeSkew bugfix (resampling corrected)
- FixTimeSkew can be configured to run as a image-magnitude interpolator:
  run after InverseFFT and set paramter data_space=imspace in the oplist
- ComputeFieldMap bugfix (2-volume asems computes correctly)
- ComputeFieldMap and GeometricUndistortionK both changed to use single
  "fmap_file" parameter (got rid of "mask_file"). Both the fieldmap volume
  and the mask volume are contained in the single file.
- bugfix to the volrange option, starting at volume 0 or 1 works now

------------------------------------------------------------------------------
recon-tools-0.3r4     (beta)                                        9/20/2006
- added left-out "lmbda" parameter in GeometricUndistortionK
- FixTimeSkew works with 2shot data

root/recon/analyze.py
"This module implements details of the Analyze7.5 file format."
import numpy as np
import struct
import sys
from recon.odict import odict
from recon import struct_unpack, struct_pack, NATIVE, LITTLE_ENDIAN, BIG_ENDIAN
from recon.util import Quaternion, range_exceeds, integer_ranges
from recon.imageio import ReconImage

# datatype is a bit flag into the datatype identification byte of the Analyze
# header. 
UBYTE = 2
SHORT = 4
INTEGER = 8
FLOAT = 16
COMPLEX = 32
DOUBLE = 64

# map Analyze datatype to numpy dtype
datatype2dtype = {
  UBYTE: np.dtype(np.uint8),
  SHORT: np.dtype(np.int16),
  INTEGER: np.dtype(np.int32),
  FLOAT: np.dtype(np.float32),
  DOUBLE: np.dtype(np.float64),
  COMPLEX: np.dtype(np.complex64)}

# map numpy dtype to Analyze datatype
dtype2datatype = \
  dict([(v,k) for k,v in datatype2dtype.items()])


datatype2bitpix = dict([(dc, dt.itemsize*8)
                        for dc,dt in datatype2dtype.items()])

orientname2orientcode = {
    "radiological": 0,
    "coronal": 1,
    "sagittal": 2,
    "transverse": 3,
    "coronal_flipped": 4,
    "sagittal_flipped": 5,
    }

orientcode2orientname = dict([(v,k) for k,v in orientname2orientcode.items()])

# these are supposed to xform the current image into [R,A,S]
# such that M*[X,Y,Z]^T = [R,A,S]^T
# this is opposite the ANALYZE system in handedness
xforms = {
    "radiological": np.array([[-1., 0., 0.],
                              [ 0., 1., 0.],
                              [ 0., 0., 1.],]),
    "transverse": np.array([[-1., 0., 0.],
                            [ 0.,-1., 0.],
                            [ 0., 0., 1.],]),
    "coronal": np.array([[-1., 0., 0.],
                         [ 0., 0., 1.],
                         [ 0., 1., 0.],]),
    "coronal_flipped": np.array([[-1., 0., 0.],
                                 [ 0., 0., 1.],
                                 [ 0.,-1., 0.]],),
    
    "sagittal": np.array([[ 0., 0.,-1.],
                          [ 1., 0., 0.],
                          [ 0., 1., 0.],]),
    
    "sagittal_flipped": np.array([[ 0., 0.,-1.],
                                  [ 1., 0., 0.],
                                  [ 0.,-1., 0.],]),

    }

def canonical_orient(xform):
    for k,v in xforms.items():
        if np.allclose(xform, v, atol=1e-6):
            return k
    return ""

# what is native and what is swapped?
byteorders = {
    sys.byteorder: sys.byteorder=='little' and LITTLE_ENDIAN or BIG_ENDIAN,
    'swapped': sys.byteorder=='little' and BIG_ENDIAN or LITTLE_ENDIAN
    }

HEADER_SIZE = 348
struct_fields = odict((
    ('sizeof_hdr','i'),
    ('data_type','10s'),
    ('db_name','18s'),
    ('extents','i'),
    ('session_error','h'),
    ('regular','c'),
    ('hkey_un0','c'),
    ('ndim','h'),
    ('idim','h'),
    ('jdim','h'),
    ('kdim','h'),
    ('tdim','h'),
    ('dim5','h'),
    ('dim6','h'),
    ('dim7','h'),
    ('vox_units','4s'),
    ('cal_units','8s'),
    ('unused1','h'),
    ('datatype','h'),
    ('bitpix','h'),
    ('dim_un0','h'),
    ('pixdim0','f'),
    ('isize','f'),
    ('jsize','f'),
    ('ksize','f'),
    ('tsize','f'),
    ('pixdim5','f'),
    ('pixdim6','f'),
    ('pixdim7','f'),
    ('vox_offset','f'),
    ('scale_factor','f'),
    ('funused2','f'),
    ('funused3','f'),
    ('cal_max','f'),
    ('cal_min','f'),
    ('compressed','i'),
    ('verified','i'),
    ('glmax','i'),
    ('glmin','i'),
    ('descrip','80s'),
    ('aux_file','24s'),
    ('orient','B'),
    ('x0','h'),
    ('y0','h'),
    ('z0','h'),
    ('sunused','4s'),
    ('generated','10s'),
    ('scannum','10s'),
    ('patient_id','10s'),
    ('exp_date','10s'),
    ('exp_time','10s'),
    ('hist_un0','3s'),
    ('views','i'),
    ('vols_added','i'),
    ('start_field','i'),
    ('field_skip','i'),
    ('omax','i'),
    ('omin','i'),
    ('smax','i'),
    ('smin','i')))

field_formats = struct_fields.values()


##############################################################################
class AnalyzeReader:
    """
    Loads image data from an ANALYZE 7.5 file.
    """

    #-------------------------------------------------------------------------
    def __init__(self, filestem, target_dtype=None, vrange=()):
        self.load_header(filestem+".hdr")
        self.load_image(filestem+".img", target_dtype, vrange)

    #-------------------------------------------------------------------------
    def load_header(self, filename):
        "Load Analyze7.5 header from the given filename"

        # Determine byte order of the header.  The first header element is the
        # header size.  It should always be 384.  If it is not then you know
        # you read it in the wrong byte order.

        byte_order = byteorders[sys.byteorder]
        self.swapped = False
        reported_length = struct_unpack(file(filename, 'rb'),
          byte_order, field_formats[0:1])[0]
        if reported_length != HEADER_SIZE:
            byte_order = byteorders['swapped']
            self.swapped=True

        # unpack all header values
        values = struct_unpack(file(filename, 'rb'), byte_order, field_formats)

        # now load values into self
        hd_vals = dict()
        map(hd_vals.__setitem__, struct_fields.keys(), values)
        self.__dict__.update(hd_vals)
    #-------------------------------------------------------------------------
    def load_image(self, filename, target_dtype, vrange):
        # bytes per pixel
        bytepix = self.bitpix/8
        numtype = datatype2dtype[self.datatype]
        byteoffset = 0
        if target_dtype is not None and \
               not range_exceeds(target_dtype, numtype):
            raise ValueError("the dynamic range of the desired datatype does "\
                             "not exceed that of the raw data")            
        # need to cook tdim if vrange is set
        if self.tdim and vrange:
            vend = (vrange[1]<0 or vrange[1]>=self.tdim) \
                   and self.tdim-1 or vrange[1]
            vstart = (vrange[0] > vend) and vend or vrange[0]
            self.tdim = vend-vstart+1
            byteoffset = vstart*bytepix*np.product((self.kdim,
                                                    self.jdim,self.idim))

        dims = self.tdim > 1 and (self.tdim, self.kdim, self.jdim, self.idim) \
               or (self.kdim, self.jdim, self.idim)
        datasize = bytepix * np.product(dims)
        fp = file(filename, 'rb')
        fp.seek(byteoffset, 1)
        data = np.fromstring(fp.read(datasize),numtype).reshape(dims)
        if self.swapped: data = data.byteswap()
        if target_dtype is not None and target_dtype != numtype:
            if target_dtype not in integer_ranges.keys():
                scale = self.scale_factor or 1.0
                self.data = (data*scale).astype(target_dtype)
            else:
                self.data = data.astype(target_dtype)
        else:
            self.data = data
        fp.close()
    #-------------------------------------------------------------------------
    def toImage(self):
        orient_name = orientcode2orientname.get(self.orient, "")
        M = xforms.get(orient_name, np.identity(3))
        quat = Quaternion(M=M)
        offset_ana = np.array([self.x0*self.isize, self.y0*self.jsize,
                               self.z0*self.ksize])
        offset = -np.dot(M, offset_ana)
        dimlengths = np.array([self.ksize, self.jsize, self.isize]) * \
                     np.array(self.data.shape[-3:])
##         # sanity check: if the offset values were funky, don't try to set it
##         if not (np.abs(offset*2) < dimlengths).all():
##             offset = None
        return ReconImage(self.data.copy(), self.isize, self.jsize,
                          self.ksize, self.tsize, offset=offset,
                          scaling=(self.scale_factor or 1.0),
                          orient_xform=quat)

##############################################################################
class AnalyzeWriter (object):
    """
    Write a given image into a single Analyze7.5 format hdr/img pair.
    """

    #[STATIC]
    _defaults_for_fieldname = {
      'sizeof_hdr': HEADER_SIZE,
      'extents': 16384,
      'regular': 'r',
      'hkey_un0': ' ',
      'vox_units': 'mm',
      'scale_factor':1.}
    #[STATIC]
    _defaults_for_descriptor = {'i': 0, 'h': 0, 'f': 0., 'c': '\0', 's': ''}

    #-------------------------------------------------------------------------
    def __init__(self, image, dtype=None, scale=1.0):
        self.image = image
        self.scaling = scale
        if dtype is None:
            dtype = image[:].dtype
        self.datatype = dtype2datatype.get(dtype, None)
        if not self.datatype:
            raise ValueError("This data type is not supported by the "\
                             "ANALYZE format: %s"%dtype)
        self._dataview = _construct_dataview(image[:].dtype,
                                             dtype, self.scaling)
    #-------------------------------------------------------------------------
    def _default_field_value(self, fieldname, fieldformat):
        "[STATIC] Get the default value for the given field."
        return self._defaults_for_fieldname.get(fieldname, None) or \
               self._defaults_for_descriptor[fieldformat[-1]]
    #-------------------------------------------------------------------------
    def write(self, filestem):
        "Write ANALYZE format header, image file pair."
        headername, imagename = "%s.hdr"%filestem, "%s.img"%filestem
        self.write_hdr(headername)
        self.write_img(imagename)

    #-------------------------------------------------------------------------
    def write_hdr(self, filename):
        "Write ANALYZE format header (.hdr) file."
        image = self.image

        # NOT IMPLEMENTING YET
        # ReconImage r0 is the offset in xyz-space, but ANALYZE
        # calls for an offset in vox-space
        xform = image.orientation_xform.tomatrix()
        orientation_name = canonical_orient(xform)
        inv_xform = np.linalg.inv(xform)
        r0 = np.array([image.x0, image.y0, image.z0])
        dimscl = np.array([image.isize, image.jsize, image.ksize])
        r0_ana = (-np.dot(inv_xform, r0)/dimscl).astype(np.int16)

        dtype = datatype2dtype[self.datatype]
        if dtype in np.sctypes['int']:
            glmax = 2**(datatype2bitpix[self.datatype]-1) - 1
        elif dtype in np.sctypes['uint']:
            glmax = 2**(datatype2bitpix[self.datatype]) - 1
        else:
            glmax = 1
        imagevalues = {
          'datatype': self.datatype,
          'bitpix': datatype2bitpix[self.datatype],
          # should be 4 for (z,y,x) and (t,z,y,x) ...
          # should be 3 for (y,x) ??
          'ndim': (image.ndim >= 3 and 4 or image.ndim),
          'idim': image.idim,
          'jdim': image.jdim,
          'kdim': image.kdim,
          'tdim': (image.tdim or 1),
          'isize': image.isize,
          'jsize': image.jsize,
          'ksize': image.ksize,
          'tsize': image.tsize,
          'x0': r0_ana[0],
          'y0': r0_ana[1],
          'z0': r0_ana[2],
          'scale_factor': self.scaling,
          # SPM says that these are (0,2^bitpix-1) for integers,
          # or (0,1) for floating points
          'glmin': 0,
          'glmax': glmax,
          'cal_min': 0.0, # don't suggest display color mapping
          'cal_max': 0.0,
          'orient': orientname2orientcode.get(orientation_name,255),
        }

        def fieldvalue(fieldname, fieldformat):
            if imagevalues.has_key(fieldname):
                return imagevalues[fieldname]
            return self._default_field_value(fieldname, fieldformat)

        fieldvalues = [fieldvalue(*field) for field in struct_fields.items()]
        header = struct_pack(NATIVE, field_formats, fieldvalues)
        file(filename,'wb').write(header)

    #-------------------------------------------------------------------------
    def write_img(self, filename):
        "Write ANALYZE format image (.img) file."
        # Write the image file.
        f = file(filename, 'wb')
        f.write(self._dataview(self.image[:]).tostring())
        f.close()

#-----------------------------------------------------------------------------
def _construct_dataview(static_dtype, new_dtype, scaling):
    xform = lambda d: d
    if static_dtype in np.sctypes['complex'] and \
       new_dtype not in np.sctypes['complex']:
        xform = lambda d: abs(d)
    if scaling != 1.0:
        # assume that this will need to be rounded to Int, so add 0.5
        xform = lambda d,g=xform: g(d)/scaling + 0.5
    xform = lambda d, g=xform: g(d).astype(new_dtype)
    return xform

#-----------------------------------------------------------------------------
def readImage(filename, **kwargs):
    return AnalyzeReader(filename, **kwargs).toImage()
#-----------------------------------------------------------------------------
def writeImage(image, filestem, **kwargs):
    AnalyzeWriter(image,**kwargs).write(filestem)
#-----------------------------------------------------------------------------
def dumpHeader(filename):

    # Determine byte order of the header.  The first header element is the
    # header size.  It should always be 384.  If it is not then you know
    # you read it in the wrong byte order.
        
    byte_order = byteorders[sys.byteorder]
    reported_length = struct_unpack(file(filename, 'rb'),
                                    byte_order, field_formats[0:1])[0]
    if reported_length != HEADER_SIZE:
        byte_order = byteorders['swapped']
            
    # unpack all header values
    values = struct_unpack(file(filename, 'rb'), byte_order, field_formats)
    
    for field,val in zip(struct_fields.keys(), values):
        print field, " -- ", val
